### ISSUE TITLE
Sorting by weighted fields

### ISSUE THREAD

**Issue by impactdni2** [typesense/typesense#692]
Working on evaluating Typesense for an upgrade to our search platform, and trying to figure out if there is a way to adjust sort values based on columns.  Maybe someone can point me in the right direction, or let me know if its something that would be appreciated as a feature request...

Essentially trying to nudge result sort_by values based on various fields.

A few examples would be things like:
- Sorting by number of likes (sort_by=likes:desc), but with a bump for having a verified profile.  Such as sort_score = (verified_profile=1 ? 100 : 0) + likes.  A result with 200 likes (no verified profile) would score/sort the same as a result with 100 likes but verified.
- Or in an example like sorting by geo distance, moving results with a photo closer by some defined amount (5 miles, etc).  For the example of 5mi, a result 3mi away but no photo would have the same score as a result 8mi away with a photo.

I know there is something similar in geo search with bucketing - which is close, but not quite the same due to the edges of buckets, and the feature-request for decay (https://github.com/typesense/typesense/issues/376)  looks pretty close, but not exactly the same.

Is there any way to do this that I missed?  Is there any interest in something like that as a feature-request?

**Comment by kishorenc** [comment]
@impactdni2 

We've just wrapped up a feature available in the `typesense/typesense:0.24.0.rc29` Docker build which allows you to sort documents based on the result of a filter expression. Let's say you want to boost products whose price is above $100 and also whose color is red. You can do:

```
"sort_by": "_eval(price:>100 && color:red):desc,_text_match:desc"
```

The above sort_by clause will sort all red colored, >$100 priced products ahead of other products. So, while it doesn't support a ternary operator like your example, you can evaluate a filter expression to a boolean value and sort on that. Let me know if that helps!

**Comment by impactdni2** [comment]
Thanks a ton for the reply, really appreciate you taking the time to look at it :)

I'll give the _eval() option a try, I think its pretty close to what we're looking for, but I think we'll still see the issue we're having unfortunately.  The main reason for a boost is specifically to make the results look a bit more "natural".  If we use the likes/verified example from above, I think it illustrates the issue.  With _eval(), a user would basically see all verified users right away, then all the result of the "natural" results.  Where with a boost, you'd see the verified results show with some favorability, but would still see unverified if they were better (more likes).

**Comment by kishorenc** [comment]
Yeah, this requires us to allow an expression to compute a score in the sort_by clause. Will consider as a feature request.

**Comment by TroyKomodo** [comment]
Bump on this 👍

**Comment by TroyKomodo** [comment]
I think what this ticket was about is:
if we have 3 integer fields.
`all_time_popularity`
`weekly_popularity`
`_text_match`

we want to rank them by 
```
max(all_time_popularity, <max-weight>) * max(weekly_popularity, <max-weight>) * _text_match
```
or something to basically merge different scores, rather than right now we sort by 1 then the next.

**Comment by stagge-skovik** [comment]
I have a similar need. I'd like to demote archived users in search results without fully excluding them, so that a perfect match on an archived user still ranks above weak matches on active ones. 
 
 Are there any updates on when this might be implemented? Or are there any work-arounds? Would love to see this land. Thanks!

**Comment by a-mutharasu** [comment]
This is still not possible today, and it's worth saying why precisely, because the reason also
explains why bucketing and decay get close but never quite fit.

`sort_by` compares up to three `int64_t` tiers **lexicographically** — tier 1 decides, ties fall to
tier 2, then tier 3. `Index::compute_sort_scores` fills a fixed `int64_t scores[3]` and there is no
arithmetic path between the tiers. So `sort_score = (verified_profile ? 100 : 0) + likes` cannot be
expressed: "verified" and "likes" can each be a tier, but they can't be *added*. Bucketing
approximates it precisely because it flattens one signal into tie groups so the next tier can act as
a tiebreak — hence the edge effects you mentioned.

A single arithmetic sort clause would cover both of your examples directly:

```
sort_by=_score(likes + 100 * _eval([(verified_profile:=true):1])):desc
```

and the geo one, nudging results with a photo 5 miles closer:

```
sort_by=_score(loc(lat, lng) - 8047 * _eval([(has_photo:=true):1])):asc
```

I've written up the full design — grammar, cost, and the one structural obstacle — as a comment on
#1976, which is asking for the same capability from the currency-conversion angle. Rather than
duplicate it here: https://github.com/typesense/typesense/issues/1976#issuecomment-5155849499

Short version of the obstacle: `compute_sort_scores` gets one sort-index pointer per tier
(`std::array<..., 3> field_values`), so an expression touching more numeric fields than there are
tiers has nowhere to read them from. That's the part worth agreeing on before anyone writes it.

Separately, I've sent a PR for a smaller self-contained piece — #3011 (`_eval_sum`, closing #2014),
so that a list of additive boosts actually sums instead of first-match-wins. It doesn't solve this issue on its own
(it still can't add a numeric field into the same tier), but it's the primitive the above builds on.
